### DISCUSSION TITLE
Should strategy_saturation_largest_first() work when called with colors != {}

### DISCUSSION THREAD

**Discussion by gpdwatkins** [networkx/networkx#5884] [Ideas]
The docstring for strategy_saturation_largest_first() function says that  

```
``colors`` is a dictionary mapping nodes of ``G`` to colors, for those nodes that have already been colored. 
```

However it only works if the generator is initialised with colors={}

Currently it seems like the function is only called by greedy_color, with colors={}. In that case, it works. However I wanted to use DSATUR to finish coloring a part-colored graph, so I called strategy_saturation_largest_first() directly. I had to change the function so that it would accept a partial coloring.

Is this a functionality that is desired? It feels like it should be (especially given the docstring) - I can submit a pull request to add this functionality?

**Comment by dschult** [comment]
The doc_strings in that module are not very complete. There is one commit when they got expanded from single lines to about 5-10 lines. But I can believe that they are not aligned with what the code is doing. 

I believe that you are correct that this function would be improved and match the doc_string better by allowing the input `colors` dict to be partially filled. It would also please me to see the case `i==0` pulled out of that loop.  I think you will need to do that anyway when colors is not an empty dict.

It would be great to have a pull request to improve this function -- and add some tests that assure it stays working that way. 
Thanks!

**Reply by gpdwatkins** [reply]
Thanks @dschult - I've adapted the function so it does what the docstring suggests, and submitted a pull request with a test that checks it's working.
